Zend/zend_ast.c

```diff
@@ -441,3 +444,1011 @@ ZEND_API void zend_ast_apply(zend_ast *ast, zend_ast_apply_func fn) {
+	for (i = 0; i < s->len; i++) {
+		unsigned char c = s->val[i];
+		if (c == '\'' || c == '\\') {
+			smart_str_appendc(str, '\\');
+			smart_str_appendc(str, c);
+		} else {
+			smart_str_appendc(str, c);
+		}
+	}
+}
+
+static void zend_ast_export_dstr(smart_str *str, zend_string *s)
+{
+	size_t i;
+
+	for (i = 0; i < s->len; i++) {
+		unsigned char c = s->val[i];
+		if (c < ' ') {
+			switch (c) {
+				case '\n':
+					smart_str_appends(str, "\\n");
+					break;
+				case '\r':
+					smart_str_appends(str, "\\r");
+					break;
+				case '\t':
+					smart_str_appends(str, "\\t");
+					break;
+				case '\f':
+					smart_str_appends(str, "\\f");
+					break;
+				case '\v':
+					smart_str_appends(str, "\\v");
+					break;
+#ifdef PHP_WIN32
+				case VK_ESCAPE:
+#else
+				case '\e':
+#endif
+					smart_str_appends(str, "\\e");
+					break;
+				default:
+					smart_str_appends(str, "\\0");
+					smart_str_appendc(str, '0' + (c / 8));
+					smart_str_appendc(str, '0' + (c % 8));
+					break;
+			}
+		} else {
+			if (c == '"' || c == '`' || c == '$' || c == '\\') {
+				smart_str_appendc(str, '\\');
+			}
+			smart_str_appendc(str, c);
+		}
+	}
+}
+
+static void zend_ast_export_indent(smart_str *str, int indent)
+{
+	while (indent > 0) {
+		smart_str_appendc(str, '\t');
+		indent--;
+	}
+}
+
+static void zend_ast_export_name(smart_str *str, zend_ast *ast, int priority, int indent)
+{
+	if (ast->kind == ZEND_AST_ZVAL) {
+		zval *zv = zend_ast_get_zval(ast);
+
+		if (Z_TYPE_P(zv) == IS_STRING) {
+			smart_str_appendl(str, Z_STRVAL_P(zv), Z_STRLEN_P(zv));
+			return;
+		}
+	}
+	zend_ast_export_ex(str, ast, priority, indent);
+}
+
+static void zend_ast_export_var(smart_str *str, zend_ast *ast, int priority, int indent)
+{
+	if (ast->kind == ZEND_AST_ZVAL) {
+		zval *zv = zend_ast_get_zval(ast);
+		if (Z_TYPE_P(zv) == IS_STRING) {
+			smart_str_appendl(str, Z_STRVAL_P(zv), Z_STRLEN_P(zv));
+			return;
+		}
+	} else if (ast->kind == ZEND_AST_VAR) {
+		zend_ast_export_ex(str, ast, 0, indent);
+		return;
+	}
+	smart_str_appendc(str, '{');
+	zend_ast_export_name(str, ast, 0, indent);
+	smart_str_appendc(str, '}');
+}
+
+static void zend_ast_export_list(smart_str *str, zend_ast_list *list, int separator, int priority, int indent)
+{
+	uint32_t i = 0;
+
+	while (i < list->children) {
+		if (i != 0 && separator) {
+			smart_str_appends(str, ", ");
+		}
+		zend_ast_export_ex(str, list->child[i], priority, indent);
+		i++;
+	}
+}
+
+static void zend_ast_export_encaps_list(smart_str *str, zend_ast_list *list, int indent)
+{
+	uint32_t i = 0;
+	zend_ast *ast;
+
+	while (i < list->children) {
+		ast = list->child[i];
+		if (ast->kind == ZEND_AST_ZVAL) {
+			zval *zv = zend_ast_get_zval(ast);
+
+			if (Z_TYPE_P(zv) == IS_STRING) {
+				zend_ast_export_dstr(str, Z_STR_P(zv));
+			} else {
+				zend_ast_export_ex(str, ast, 0, indent);
+			}
+		} else if (ast->kind == ZEND_AST_VAR && ast->child[0]->kind == ZEND_AST_ZVAL) {
+			zend_ast_export_ex(str, ast, 0, indent);
```

Comment:
This is unsafe in the following case: `"{$a}b"`. This would be pretty printed as `"$ab"`, with different semantics.